### ISSUE TITLE
Cmd+A doesn't select all on Mac

### ISSUE THREAD

**Issue by JoelEinbinder** [puppeteer/puppeteer#1313]
There are two types of shortcuts: those handled by OS and those handled by Chrome
  - `Alt-Tab` is an example of a **native** shortcut
  - `Shift-Left` is an example of **Chrome-handled** shortcut

Puppeteer doesn't emulate native shortcuts because native shortcuts depend on the active window, which is out of control for puppeteer.

Certain shortcuts, such as "select all", are handled differently on different platforms:
- On Win and Linux, `Ctrl-A` is handled by Chrome, and puppeteer **does** emulate this
- On OS X, `⌘-A` is handled by OS X, and puppeteer **doesn't** emulate this

**Comment by drorbiran** [comment]
I'm also having an issue with simulating a paste from clipboard.
tried 

```
 await this.page.keyboard.down('ControlLeft');
 await this.page.keyboard.down('V');

 await this.page.keyboard.up('V');
  await this.page.keyboard.up('ControlLeft');
```

**Comment by dragoroff** [comment]
You can solve your problem by clicking 3 times in a row and then deleting your text.
` await input.click({ clickCount: 3 });
   await page.keyboard.press("Backspace");`

**Comment by aadityataparia** [comment]
I'm also having an issue with simulating command click:

```js
 await page.keyboard.down('Meta');
 await page.click('a');
 await page.keyboard.up('Meta');
```

It doesn't open a new tab.

**Comment by swissspidy** [comment]
Why exactly is this issue closed? 🤔

**Comment by voshond** [comment]
Yes, I'm wondering too.

Unfortunately in some special cases the triple click does not select all the text, in my case the website has some weird handler where i cannot simply put the data as the value for the input box and then send, i actually have to type the input and before delete the original, probably due to some eventhandler listing to keyboard input.

**Comment by aduth** [comment]
This is still a problem in my testing. I'd think the issue should be reopened.

In the meantime, what's worked for me as a viable alternative for "select all" is:

```js
await page.evaluate( () => document.execCommand( 'selectall', false, null ) );
```

**Comment by swissspidy** [comment]
@JoelEinbinder @aslushnikov Can we reopen this?

**Comment by aslushnikov** [comment]
@swissspidy it might be a little unclear, but the issue is *stating* the state of the world that's beyond our reach. The issue is closed since it's not actionable - the best we can do is to put this in our [docs](https://pptr.dev/#?product=Puppeteer&version=v1.13.0&show=api-class-keyboard).

The workaround with `selectall` should be good-enough for the most usecases.

**Comment by aduth** [comment]
> The workaround with `selectall` should be good-enough for the most usecases.

Expanding on my previous comment, it ultimately wasn't quite enough for our application which included more advanced handling to `preventDefault` the Cmd+A behavior under some conditions.

The full implementation ended up being just a bit more complex:

```js
/**
 * Emulates a Ctrl+A SelectAll key combination by dispatching custom keyboard
 * events and using the results of those events to determine whether to call
 * `document.execCommand( 'selectall' );`. This is necessary because Puppeteer
 * does not emulate Ctrl+A SelectAll in macOS. Events are dispatched to ensure
 * that any `Event#preventDefault` which would have normally occurred in the
 * application as a result of Ctrl+A is respected.
 *
 * @link https://github.com/GoogleChrome/puppeteer/issues/1313
 * @link https://w3c.github.io/uievents/tools/key-event-viewer.html
 *
 * @return {Promise} Promise resolving once the SelectAll emulation completes.
 */
async function emulateSelectAll() {
	await page.evaluate( () => {
		const isMac = /Mac|iPod|iPhone|iPad/.test( window.navigator.platform );

		document.activeElement.dispatchEvent(
			new KeyboardEvent( 'keydown', {
				bubbles: true,
				cancelable: true,
				key: isMac ? 'Meta' : 'Control',
				code: isMac ? 'MetaLeft' : 'ControlLeft',
				location: window.KeyboardEvent.DOM_KEY_LOCATION_LEFT,
				getModifierState: ( keyArg ) => keyArg === ( isMac ? 'Meta' : 'Control' ),
				ctrlKey: ! isMac,
				metaKey: isMac,
				charCode: 0,
				keyCode: isMac ? 93 : 17,
				which: isMac ? 93 : 17,
			} )
		);

		const preventableEvent = new KeyboardEvent( 'keydown', {
			bubbles: true,
			cancelable: true,
			key: 'a',
			code: 'KeyA',
			location: window.KeyboardEvent.DOM_KEY_LOCATION_STANDARD,
			getModifierState: ( keyArg ) => keyArg === ( isMac ? 'Meta' : 'Control' ),
			ctrlKey: ! isMac,
			metaKey: isMac,
			charCode: 0,
			keyCode: 65,
			which: 65,
		} );

		const wasPrevented = (
			! document.activeElement.dispatchEvent( preventableEvent ) ||
			preventableEvent.defaultPrevented
		);

		if ( ! wasPrevented ) {
			document.execCommand( 'selectall', false, null );
		}

		document.activeElement.dispatchEvent(
			new KeyboardEvent( 'keyup', {
				bubbles: true,
				cancelable: true,
				key: isMac ? 'Meta' : 'Control',
				code: isMac ? 'MetaLeft' : 'ControlLeft',
				location: window.KeyboardEvent.DOM_KEY_LOCATION_LEFT,
				getModifierState: () => false,
				charCode: 0,
				keyCode: isMac ? 93 : 17,
				which: isMac ? 93 : 17,
			} ),
		);
	} );
}
```

**Comment by ZAYDEK** [comment]
Just to follow up, I was struggling with this too. I found a StackOverflow that asked and answered the same question so I wanted to repost the relevant snippet here

```
await page.keyboard.down("Control")
await page.keyboard.press("A")
await page.keyboard.up("Control")
```

This surprised me because I’m on macOS but control still worked. OP here: https://stackoverflow.com/a/52633857.

**Comment by Jiang-Xuan** [comment]
`Command + v` is not working on mac also, but i find a solution, https://stackoverflow.com/questions/11750447/performing-a-copy-and-paste-with-selenium-2#answer-41046276 . in my mac, i confirmed that it works.

**Comment by avalanche1** [comment]
What about
`document.execCommand("selectAll");`
?
maybe that won't work in headless mode...

**Comment by CurtisMorice** [comment]
We are using Backbone.js so my code is a little different but I made it work with 

    {method: 'click',  value: '[tg-name=passcode_input]'},
    {method: ['keyboard', 'down'],  value: 'Alt'},
    {method: ['keyboard', 'press'], value: 'ArrowRight'},
    {method: ['keyboard', 'down'],  value: 'Alt'},
    {method: ['keyboard', 'down'],  value: 'Shift'},
    {method: ['keyboard', 'press'], value: 'ArrowLeft'},
    {method: ['keyboard', 'up'],    value: 'Alt'},
    {method: ['keyboard', 'up'],    value: 'Shift'},
    {method: ['keyboard', 'press'], value: 'Backspace'},

basically pressing/click into the input , then command (Alt) + ArrowRight,  then command (Alt) + Shift + ArrowLeft, and then Backspace/Delete

Hope this is understood and works for you (anyone?)

**Comment by eeeeaaii** [comment]
I get it that you can't hijack Alt on Linux/Windows machines but you CAN do this on a Mac (it's the Option key on a mac).

Likewise, on a Mac you can easily create browser behavior that uses the Option key, javascript that recognizes when the Option key is pressed, etc.

I feel like it would make more sense to have this be platform-dependent, right now the implemention on Mac is limited by windows/Linux compatibility issues that shouldn't affect it.

Even in the name of this bug it says "on the mac" so I'm not sure why it's closed, the problems cited for reasons for closing this bug aren't really relevant to the Mac.

**Comment by zeshhaan** [comment]
> ```js
> ;
> ```

Can we have same kind of command for paste too?
`await page.evaluate( () => document.execCommand( 'paste'));`

**Comment by manatarms** [comment]
I need to do `⌘ + i` on macOS. Has anyone been able to figure out how to trigger the meta shortcuts in macOS?

**Comment by GouravSingh2580** [comment]
How can i cut cut with`page.keyboard.down("Control")` on mac? `page.keyboard.down("X")` after that doesn't work

**Comment by wilsonssss** [comment]
> You can solve your problem by clicking 3 times in a row and then deleting your text. ` await input.click({ clickCount: 3 }); await page.keyboard.press("Backspace");`

This works for me 
`
await elementHandle.click({ count: 3 })
await page.keyboard.press('Backspace') 
`

**Comment by richoandika** [comment]
```
if (!page.$(selector)) {
    return;
}
await page.click(selector, { clickCount: 3 });
await page.type(selector, '');
```
